sstables/kl/reader_impl.hh

```diff
@@ -194,15 +194,9 @@ private:
         while (true) {
             if (_state == state::ROW_START) {
                 _state = state::NOT_CLOSING;
-                if (read_short_length_bytes(*_processing_data, _key) != read_status::ready) {
-                    co_yield row_consumer::proceed::yes;
-                }
-                if (read_32(*_processing_data) != read_status::ready) {
-                    co_yield row_consumer::proceed::yes;
-                }
-                if (read_64(*_processing_data) != read_status::ready) {
-                    co_yield row_consumer::proceed::yes;
-                }
+                co_yield read_short_length_bytes(*_processing_data, _key);
+                co_yield read_32(*_processing_data);
```

Comment:
I'm afraid I didn't understand this change (I know you tried to explain it in the commit message, but I didn't understand): you only want to "yield" if we couldn't finish read_32 - if we have enough data (4 bytes), you don't want to yield. But after this change, it looks like it does. I see in the commit message that you said that perhaps because of something (which I didn't understand) it doesn't really yield. But I don't understand why.

I suspect the magic has something to do with yield_value() in the very end of the patch - and you also mentioned it in the commit message - but I don't see this function called anywhere :-( Maybe I'm missing half of this patch which will appear later?

Reply:
The missing part is probably in the first patch of the series
I've paraphrased the C++ reference in the comment before the ```processing_result_generator```, and the main takeaway is that ```co_yield val``` is equivalent to ```co_await yield_value(val)```, so in every place where we use ```co_yield```, we also use ```yield_value```.
The following is also basically copy paste from the reference.
The result object of ```yield_value```, has 3 methods: `await_ready()`, `await_suspend()`, and `await_resume()`, but we only care about `await_ready()`. When we co_await such an object (and we `co_await` because `co_yield` = `co_await` + `yield_value`), before even stopping the coroutine, we first check the return value of its `await_ready()` method. If it returns `true` (so in our case if we yielded `read_status::ready`), we continue the coroutine without suspending it.
If it returns `false` (so in our case if we yielded `read_status::waiting`), we basically stop the execution and return control to the caller, and because of how we implemented `yield_value`, `proceed::yes` is saved in the `processing_result_generator`

If we were to compare the old and new version:
- if we got read_status::ready:
  - in the original code, the condition in the `if` is `false`, so we simply continue execution
  - in the new code, the condition in `yield_value` is `false`, so we don't store a value, and `await_ready()` returns `true`, so we continue execution
- if we got read_status::waiting:
  - in the original code, the condition in the `if` is `true`, so we stop the coroutine and yield `proceed::yes`
  - in the new code, the condition in `yield_value` is `true`, so we also save proceed::yes, and `await_ready()` returns `false`, so we also suspend execution